src/extra/utils/functions.py

```diff
@@ -45,20 +48,33 @@ def fit_gaussian(ydata, xdata=None, p0=None, norm=False, A_sign=0, **kwargs):
 
     Args:
         ydata (array_like): The data to fit. NaN's and infs will automatically
-            be masked before fitting.
-        xdata (array_like): Optional x-values corresponding to `ydata`.
+            be masked before fitting. If a [DataArray][xarray.DataArray] is
+            passed the underlying numpy array will be used.
+        xdata (array_like): Optional x-values corresponding to `ydata`. If a
+            [DataArray][xarray.DataArray] is passed the underlying numpy array
+            will be used.
         p0 (list): A list of `[y0, A, μ, σ]` to match the arguments to
             [gaussian()][extra.utils.gaussian].
         norm (bool): Whether to fit a normalized or unnormalized Gaussian.
         A_sign (int): Sign of the amplitude (A) parameter for the Gaussian.
             1 for an upwards peak, -1 for downwards. 0 (default) allows either,
             using a faster algorithm. Passing `bounds=` overrides this.
+        nans_on_failure (bool): If `True` the function will return an array of
+            NaNs of the same length as `p0`. This is useful when e.g. fitting in
+            a loop and saving the output.
         **kwargs (): All other keyword arguments will be passed to
             [curve_fit()][scipy.optimize.curve_fit].
     """
     if xdata is None:
         xdata = np.arange(len(ydata))
 
+    if _isinstance_no_import(ydata, "xarray", "DataArray"):
+        ydata = ydata.values
+    if _isinstance_no_import(xdata, "xarray", "DataArray"):
+        xdata = xdata.values
+
+    failure_value = np.full(len(p0), np.nan) if nans_on_failure else None
```

Comment:
If p0 is passed, it has to have length 4, right? Perhaps we want to check it and throw an error if not, but otherwise I don't think passing a different length list should mean a change in the return value.

```suggestion
    failure_value = np.full(4, np.nan) if nans_on_failure else None
```